ONSdigital/rm-collection-exercise-service
# add SDS entity, repo and service to add to database

# What and why?
[SDS Docs](https://github.com/ONSdigital/ras-rm-documentation/blob/main/spike/supplementary-data-service/README.md)

# How to test?
Setup
1. Deploy to your dev env
2.Once the deployment has completed, check in the DB under the 'collectionexercise' schema that the table 'supplementarydataset' has been completed
3. Once deployed, create a CE in rOps - you only need to add a period ID, the rest can be blank
4. Update this JSON with the period ID and survey ID of the CE that you have created:   
```
{
    "survey_id": [your survey ID],
    "period_id": [your period ID],
    "form_types": [
       "0017",
       "0123",
       "0001"
      ],
    "title": "A dataset version 4 for survey 2012 period 2013",
    "sds_published_at": "2023-07-17T14:46:36Z",
    "total_reporting_units": 2,
    "schema_version": "v1.0.0",
    "sds_dataset_version": 4,
    "filename": "373d9a77-2ee5-4c1f-a6dd-8d07b0ea9793.json",
    "dataset_id": "b9a87999-fcc0-4085-979f-06390fb5dddd"
  }
```
5. Goto PubSub and find the topic 'supplementary-data-service-[your-namespace]'
6. Click on the 'Message' tab in your PubSub topic and then click the 'Publish message' button

Test save initial message
7. In the message body of the pop up screen, paste the above JSON and click 'Publish'
8. Check the table 'supplementarydataset' to see that the message has been saved
9. Check the logs to see the log entry 'Successfully saved the supplementary dataset to the database'

Test updated message
11. Repeat steps 6 and 7 but alter the JSON to have a different/additional form types
12. Check the table to see that the old record has been deleted and replaced by the new message
13. Check the logs and they should have logged 'A supplementary dataset with exerciseFk has been found.', 'Supplementary dataset has been removed.' and 'Successfully saved the supplementary dataset to the database'

# Jira
https://jira.ons.gov.uk/browse/RAS-844
